README_cmake.md

```diff
@@ -304,17 +695,219 @@ and the `USE_RUNTIME` option is ignored. Other options work as expected.
+variable allows one to specify a command _prefix_ to run a target-system binary
+on the host machine. One could set this to a custom shell script that uploads
+the generator executable, runs it on the device and copies back the results.
+
+### Bypass CMake
+
+The previous two options ensure that the targets generated by
+`add_halide_library` will be _normal_ static libraries. This approach does not
+use ExternalProject, but instead produces `IMPORTED` targets. The main drawback
+of `IMPORTED` targets is that they are considered second-class in CMake. In
+particular, they cannot be installed with the typical `install(TARGETS)`
+command. Instead, they must be installed using `install(FILES)` and the
+`$<TARGET_FILE:>` generator expression.
+
+# Contributing CMake code to Halide
+
+When contributing new CMake code to Halide, keep in mind that the minimum
+version is 3.16. Therefore, it is possible (and indeed required) to use modern
+CMake best practices.
+
+Like any large and complex system with a dedication to preserving backwards
+compatibility, CMake is difficult to learn and full of traps. While not
+comprehensive, the following serves as a guide for writing quality CMake code
+and outlines the code quality expectations we have as they apply to CMake.
+
+## General guidelines and best practices
+
+The following are some common mistakes that lead to subtly broken builds.
+
+- **Reading the build directory.** While setting up the build, the build
+  directory should be considered _write only_. Using the build directory as a
+  read/write temporary directory is acceptable as long as all temp files are
+  cleaned up by the end of configuration.
+- **Prefer generator expressions.** Declarative is better than imperative and
+  this is no exception. Conditionally adding to a target property can leak
+  unwanted details about the build environment into packages. Some information
+  is not accurate or available except via generator expressions, eg. the build
+  configuration.
+- **Using the wrong variable.** `CMAKE_SOURCE_DIR` doesn't always point to the
+  Halide source root. When someone uses Halide via FetchContent, it will point
+  to _their_ source root instead. The correct variable is `Halide_SOURCE_DIR`.
+  If you want to know if the compiler is MSVC, check it directly with the `MSVC`
+  variable; don't use `WIN32`. That will be wrong when compiling with clang on
+  Windows.
+- **Using directory properties.** Directory properties have vexing behavior and
+  are essentially deprecated from CMake 3.0+. Propagating target properties is
+  the way of the future.
+- **Using the wrong visibility.** Target properties can be `PRIVATE`,
+  `INTERFACE`, or both (aka `PUBLIC`). Pick the most conservative one for each
+  scenario.
+
+### Prohibited commands list
+
+As mentioned above, using directory properties is brittle and they are therefore
+_not allowed_. The following functions may not appear in any new CMake code.
```

Comment:
...are we using any of the forbidden functions in existing code? (if so, maybe let's open a tracking issue to nuke them)

Reply:
We are not, except in three places where it is _absolutely_ necessary:

1. We use `define_property` to track the Halide targets a runtime library needs to support, which might change throughout the configure step. The property is used in a generator expression to construct a correct command line.
2. We use a `macro` in our package config code to gracefully fail by calling `return()` in the package config scope. This is required by the `find_package` mechanism. The alternative would be to duplicate a 20-line block 8 times.
3. We use `find(GLOB ...)` in our package config code to load dependency scripts that might or might not be there depending on platform/configuration. It's safe here because it's not globbing for source files for a target, but for CMake scripts to be included.